torch/_inductor/lowering.py

```diff
@@ -4386,66 +4384,67 @@ def _avg_poolnd(
-            padding,
-            ceil_mode,
-            count_include_pad,
-            divisor_override,
-        )
-
-    def fn_sum(idx, loader):
+    def fn_inner(idx, reduction_idx):
         prefix = idx[:-dim]
-        b = idx[-dim:]
-        total = None
-        for ih in itertools.product(*[range(kernel_size[i]) for i in range(dim)]):
-            inp = [b[i] * stride[i] + ih[i] - padding[i] for i in range(dim)]
-            val = loader([*prefix, *inp])
-            if total is None:
-                total = val
-            else:
-                total = ops.add(val, total)
-        return total
+        bh = idx[-dim:]
+        ih = reduction_idx
+        ih = [bh[i] * stride[i] + ih[i] - padding[i] for i in range(dim)]
+        return x_loader([*prefix, *ih])
+
+    rv = Reduction.create(
+        reduction_type="sum",
+        input_node=x,
+        device=x.get_device(),
+        dst_dtype=output_dtype,
+        src_dtype=dtype,
+        inner_fn=fn_inner,
+        ranges=new_size,
+        reduction_ranges=kernel_size,
+    )
+    if isinstance(rv.data.data, Reduction):
+        # Only realize if reduction isn't unrolled
+        rv.realize()
 
     if not had_padding or divisor_override:
         if divisor_override:
             scale = 1 / divisor_override
         else:
-            scale = 1.0 / window_size
-
-        def fn(idx):
-            return ops.mul(fn_sum(idx, x_loader), ops.constant(scale, dtype))
+            scale = 1.0 / functools.reduce(operator.mul, kernel_size)
 
+        result = mul(rv, scale)
     else:
-        ones_loader = constant_boundary_condition(
-            ones_like(x),
-            0.0,
-            padding if count_include_pad else None,
-            dim=dim,
-        )
 
-        def fn(idx):
-            # TODO(jansel): optimize to do `int(x<h)` rather than `x<h?1:0`
-            return ops.truediv(fn_sum(idx, x_loader), fn_sum(idx, ones_loader))
+        def fn_count(idx):
+            prefix = idx[:-dim]
+            bh = idx[-dim:]
 
-    rv = Pointwise.create(
-        device=x.get_device(),
-        dtype=dtype,
-        inner_fn=fn,
-        ranges=new_size,
-    )
-    # TODO(jansel): should we force these to be realized?
-    return rv
+            divide_factors = []
+            for i in range(dim):
+                hstart = bh[i] * stride[i] - padding[i]
+                hend = sympy.Min(hstart + kernel_size[i], h[i] + padding[i])
+                if not count_include_pad:
+                    hstart = sympy.Max(hstart, 0)
+                    hend = sympy.Min(hend, h[i])
+                factor = ops.index_expr(hend - hstart, torch.int32)
+                divide_factors.append(factor)
+            return functools.reduce(ops.mul, divide_factors)
+
+        divide_factor = Pointwise.create(
+            device=x.get_device(),
+            dtype=dtype,
+            inner_fn=fn_count,
+            ranges=new_size,
+        )
+        result = div(rv, divide_factor)
```

Comment:
Benchmarks show a real slowdown, I confirmed that the `adv_inception_v3` slowdown is reproducible and really calls `avg_pool2d`.

https://hud.pytorch.org/benchmark/timm_models/inductor_with_cudagraphs?startTime=Sun,%2004%20Feb%202024%2013:25:02%20GMT&stopTime=Sun,%2011%20Feb%202024%2013:25:02%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/isuruf/17/head&lCommit=f610d4dc4de17fb89b9950d05c5f39a9e4ed2019&rBranch=main&rCommit=5f6b35915af98fac741998edb55aeeba4aaa3794

One interesting note is that this division is not being fused with the reduction kernel. The cause is a known issue where reductions iterate in a different order which ends up blocking subsequent pointwise fusions in the scheduler.  

Here the reduciton loop has size
```
buf191.sizes = ([128, 192, 35, 35], [3, 3])
```
but the pointwise op has
```
buf192.sizes = ([128, 35, 35, 192], [])
```
The reordering of the 192-length dimension is what blocks the fusion in this case. Should be fixed by #100331 but that's still a work in progress.

Reply:
@shunting314 a real world example where https://github.com/pytorch/pytorch/pull/100331 would help.